### PR TITLE
16572 Alteration - Edit UI change for business type alteration

### PR DESCRIPTION
**Pull request by Jxio**
*Issue #:* /bcgov/entity[16572](https://github.com/bcgov/entity/issues/16572)

*Description of changes:*
1. Enable "change" button for BC, BEN, ULC (BC already enabled)
2. Update type dropdown options
3. Suffix change
4. Done button pressed screen
UX Design: https://preview.uxpin.com/c21a28a1679da5e2b71044251e8296995e00d9c0#/pages/162390910/simulate/sitemap

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/components/common/BcRegComments.vue
Comment (seeker25) [inline src/components/common/BcRegComments.vue:255]: Can't you just create a property localConfirmArticles like like 253?

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (seeker25) [inline src/components/common/YourCompany/ChangeBusinessType.vue:274]: Don't need `protected`, please remove for these lines

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (seeker25) [inline src/components/common/YourCompany/ChangeBusinessType.vue:397]: 397 -> 400 looks very similar to what we have in the store, might be a candidate for moving over to there.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (seeker25) [inline src/components/common/YourCompany/ChangeBusinessType.vue:419]: Are there more patterns to catch than just LTD?

File: src/components/common/YourCompany/EntityName.vue
Comment (seeker25) [inline src/components/common/YourCompany/EntityName.vue:46]: This is already too long, I'd recommend converting to a computed

File: src/components/common/YourCompany/EntityName.vue
Comment (seeker25) [inline src/components/common/YourCompany/EntityName.vue:136]: Already too long, should be converted into computed

File: src/components/common/YourCompany/EntityName.vue
Comment (seeker25) [inline src/components/common/YourCompany/EntityName.vue:162]: Already too long, should be converted into computed

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (seeker25) [inline src/components/common/YourCompany/ChangeBusinessType.vue:365]: Create a concrete type for a return?

File: src/enums/entityTypeInfo.ts
Comment (seeker25) [inline src/enums/entityTypeInfo.ts:4]: Not sure if it's worth moving these to the already existing resource files?

File: src/components/common/BcRegComments.vue
Comment (seeker25) [inline src/components/common/BcRegComments.vue:3]: I think this is fine not being in the resource file, it's fairly dynamic and requires quite a bit of html

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (seeker25) [inline src/components/common/YourCompany/ChangeBusinessType.vue:366]: This is already done in the alteration view, to choose the correct resource. If you moved the business types to the resources, you wouldn't need to do this.

File: src/components/common/BcRegComments.vue
Comment (seeker25) [inline src/components/common/BcRegComments.vue:278]: Grab from resource file

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (seeker25) [reply src/components/common/YourCompany/ChangeBusinessType.vue:366]: EG. `this.getResource.entityTypeOptions`

File: tests/unit/BcRegComments.spec.ts
Comment (seeker25) [inline tests/unit/BcRegComments.spec.ts:57]: Cool - didn't know you could do this

File: src/components/common/BcRegComments.vue
Comment (Jxio) [reply src/components/common/BcRegComments.vue:255]: ok, will do, thanks!

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (Jxio) [reply src/components/common/YourCompany/ChangeBusinessType.vue:419]: just LTD, I confirmed with @ethantspitt

File: src/enums/entityTypeInfo.ts
Comment (Jxio) [reply src/enums/entityTypeInfo.ts:4]: good catch! I'll update.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (Jxio) [reply src/components/common/YourCompany/ChangeBusinessType.vue:274]: It's originally protected there before, I can just remove it, but can't use readonly cause property's value needs to be changed during lifecycle.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (Jxio) [reply src/components/common/YourCompany/ChangeBusinessType.vue:413]: it's good to learn, thanks!

Comment (seeker25) [reply src/components/common/YourCompany/ChangeBusinessType.vue]: I think it makes sense when you're using Typescript still, outside of the Vue context.. like building services etc. 

But not inside the Vue components themselves - I don't think we do much inheritance, we mostly do composition

**Comment by pwei1018** [comment]
Temporary Url for review: https://business-edit-dev--pr-516-37q7zf1c.web.app

File: src/resources/Alteration/BC.ts
Comment (seeker25) [inline src/resources/Alteration/BC.ts:20]: Awesome

**Comment by seeker25** [comment]
As discussed, we need to move the mixin to a util hopefully.

**Comment by pwei1018** [comment]
Temporary Url for review: https://business-edit-dev--pr-516-37q7zf1c.web.app

File: package.json
Comment (severinbeauvais) [inline package.json:3]: Yes, this is a significant new feature, so I agree with increasing the minor release number.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/ChangeBusinessType.vue:10]: Check this on a narrow viewport (screen). Generally, columns should stack without weird whitespace.

(Do you have screenshots of normal vs narrow?)

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/ChangeBusinessType.vue:177]: What about CCC?

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/ChangeBusinessType.vue:376]: Comment.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/ChangeBusinessType.vue:399]: Add comment indicating what you're doing (or why). This will help the next developer to look at this, and may help the upcoming upgrade from Vuetify 2 to 3.

File: src/components/common/YourCompany/EntityName.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/EntityName.vue:307]: I like these 2 getters instead of having the code in the template.

Could you add a comment to each to describe when then are True?

(Also add return types.)

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (Jxio) [reply src/components/common/YourCompany/ChangeBusinessType.vue:177]: (changeFrom) CCC not included on UX.

File: src/resources/Alteration/BC.ts
Comment (severinbeauvais) [reply src/resources/Alteration/BC.ts:20]: Agreed! Nicely done.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/ChangeBusinessType.vue:308]: If changeData is falsy then this returns an undefined.

Should this return an empty array if changeData is falsy? Or can the code that uses this handle an undefined value?

File: src/interfaces/alteration-interfaces/alteration-interface.ts
Comment (severinbeauvais) [inline src/interfaces/alteration-interfaces/alteration-interface.ts:41]: As this isn't an "alteration interface", I recommend putting it in its own file, maybe.

Also, why "SHORT_DESC"? Could this follow our usual camel case convention?

File: src/store/store.ts
Comment (severinbeauvais) [inline src/store/store.ts:1395]: Why the "if"? Is there no way to set this to False?

File: src/utils/resource-utils.ts
Comment (severinbeauvais) [inline src/utils/resource-utils.ts:1]: Could this fit in with the Resource Model that already exists? See store/state/ and store/store.ts.

File: src/components/common/BcRegEntityDetails.vue
Comment (severinbeauvais) [inline src/components/common/BcRegEntityDetails.vue:20]: This component ^^ doesn't do much. I think it would be a better design to have this component handle the showing and hiding of the content (ie, fully encapsulate the show/hide logic).

You could do this by passing a template with the content to display (ie, the help section below).

See also HelpSection.vue.

File: src/components/common/BcRegEntityDetails.vue
Comment (severinbeauvais) [inline src/components/common/BcRegEntityDetails.vue:3]: I'm sort of OK with having the HTML for all entity types in this one file, but did you consider having separate sub-components for each type?

You could use dynamic components here.
Example: https://github.com/bcgov/business-filings-ui/blob/fafe0c3af7da84fa91f3677f785cfe5c7f0adb8b/src/components/Dashboard/FilingHistoryList.vue#L46

You may have already tried that and your current implementation may be cleaner. If so, say so :)

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (seeker25) [reply src/components/common/YourCompany/ChangeBusinessType.vue:308]: Probably could write 

`this.getResource.changeData?.entityTypeOptions || []`

File: src/utils/resource-utils.ts
Comment (seeker25) [reply src/utils/resource-utils.ts:1]: It was rewritten as a mixin, I suggested putting into utils

File: src/components/common/BcRegComments.vue
Comment (severinbeauvais) [reply src/components/common/BcRegComments.vue]: Theoretically, protected would allow the template to see this property but it wouldn't be visible outside this class/component.

However that doesn't work, so I'm OK with removing the accessor when the template needs to see this property.

File: src/components/common/BcRegEntityDetails.vue
Comment (seeker25) [reply src/components/common/BcRegEntityDetails.vue:3]: Would consider it, but this is mostly a 1 off?

File: src/components/common/YourCompany/EntityName.vue
Comment (severinbeauvais) [reply src/components/common/YourCompany/EntityName.vue]: When you change this to a computed (get) then you also have the opportunity to add a comment to describe the particular business logic (which I recommend doing, if easy).

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue]: Sometimes, some methods are called externally (eg, for validation/reset).

Still, the accessors don't work correctly in TS so it's not a big deal.

**Review by severinbeauvais — COMMENTED** [review]
Looks very good 👍 

I'll review your changes then approve if OK.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue:177]: Please confirm with UX designer.

File: src/utils/resource-utils.ts
Comment (severinbeauvais) [reply src/utils/resource-utils.ts:1]: I guess it is more of a utility (with formal parameters) instead of a state/store query...

OK, resolving.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (seeker25) [reply src/components/common/YourCompany/ChangeBusinessType.vue:177]: We've confirmed with UX designer and BA and John Lane

Here's the matrix currently:

Comment (seeker25) [reply src/components/common/YourCompany/ChangeBusinessType.vue:177]: https://preview.uxpin.com/c21a28a1679da5e2b71044251e8296995e00d9c0#/pages/162402442/simulate/sitemap

Comment (seeker25) [reply src/components/common/YourCompany/ChangeBusinessType.vue:177]: Straight from Dana:

image

File: src/components/common/BcRegEntityDetails.vue
Comment (seeker25) [reply src/components/common/BcRegEntityDetails.vue:20]: HelpSection.vue is fairly specific, it probably should have been retooled to support slots

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (Jxio) [reply src/components/common/YourCompany/ChangeBusinessType.vue:10]: this is what it looks like before:
Pasted Graphic
after I added class="pa-0”: (normal)
Pasted Graphic 1
(narrow):
Pasted Graphic 2

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue:177]: Perfect, thanks.

File: src/components/common/BcRegEntityDetails.vue
Comment (severinbeauvais) [reply src/components/common/BcRegEntityDetails.vue:20]: Probably. I'm suggesting to keep this one but with slots.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue:10]: That's good. Thanks.

File: src/interfaces/alteration-interfaces/alteration-interface.ts
Comment (Jxio) [reply src/interfaces/alteration-interfaces/alteration-interface.ts]: I moved to its own file, and changed to camel case. "SHORT_DESC" is from existing type alteration code.

File: src/store/store.ts
Comment (Jxio) [reply src/store/store.ts]: yeah you're right, updated

File: src/components/common/BcRegEntityDetails.vue
Comment (Jxio) [reply src/components/common/BcRegEntityDetails.vue]: I put this section back instead of a component, maybe better not changing HelpSection.vue.

**Comment by pwei1018** [comment]
Temporary Url for review: https://business-edit-dev--pr-516-37q7zf1c.web.app

File: src/components/common/BcRegEntityDetails.vue
Comment (severinbeauvais) [reply src/components/common/BcRegEntityDetails.vue]: I liked the HelpSectionToggle component and recommend you revert it.

My comment above is that you should pass a slot to the component (which is currently your section block), and have all the show/hide logic in the component. Currently that logic is both in BcRegEntityDetails and HelpSectionToggle. You can free BcRegEntityDetails from knowing anything about whether the help is shown/hidden, which is a nice abstraction/ encapsulation/ separation of concern. For example:

In BcRegEntityDetails.vue:
```
<HelpSectionToggle>
  <template #content> <!-- you may be able to use section instead of template here -->
    <!-- your help content goes here -->
  </template>
```

In HelpSectionToggle.vue:
```
<template>
  <div class="help-section-toggle">
    <div
      class="info-text help-toggle"
      @click="isHelpVisible= !isHelpVisible"
    >
      <!-- put your v-icon and show/hide span here -->

      <slot name="content" v-if="isHelpVisible" />
</div>
</template>
```

File: src/components/Alteration/BcRegEntityDetails.vue
Comment (severinbeauvais) [inline src/components/Alteration/BcRegEntityDetails.vue:339]: Remove if not used.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue:401]: Well, yes, but I can see that from the code. The real question is what/why are you doing this? For example, "make icon taller so that it matches the text beside it" (or whatever).

File: src/interfaces/alteration-interfaces/alteration-interface.ts
Comment (severinbeauvais) [reply src/interfaces/alteration-interfaces/alteration-interface.ts]: Great, thanks!

File: src/components/common/BcRegEntityDetails.vue
Comment (Jxio) [reply src/components/common/BcRegEntityDetails.vue]: I'm thinking move it to ChangeBusinessType.vue:
```
        <BcRegEntityDetails
          :helpToggle="helpToggle"
          ...
        >
          <template #toggle>
            <div
              class="info-text help-toggle pt-2"
              @click="helpToggle = !helpToggle"
            >
              <v-icon
                class="pr-2 mt-n1"
                color="primary"
              >
                mdi-help-circle-outline
              </v-icon>
              <span v-if="!isHelpVisible">Learn More</span>
              <span v-else>Hide Learn More</span>
            </div>
          </template>
        </BcRegEntityDetails>
```
at BcRegEntityDetails.vue:
```
<!-- Help Section Toggle -->
<slot name="toggle" />
```
feel HelpSectionToggle can be removed cause it's too simple. what do you think?

Comment (severinbeauvais) [reply src/components/common/BcRegEntityDetails.vue]: Would you have multiple copies of the above?

I'm looking for simple/clean/easy-to-review code and use of sub-components as needed to accomplish that.

There are different ways to lay this out. Can I leave this with you?

Comment (Jxio) [reply src/components/common/BcRegEntityDetails.vue]: no multiple copies, thanks for that! I updated all code under  on BcRegEntityDetails with slots, I think better to keep as its own template cause it's specific to alteration, not for common components.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (Jxio) [reply src/components/common/YourCompany/ChangeBusinessType.vue]: I removed my comment also removed my change of `:deep`, cause it doesn't need anymore after using slots for "help toggle" section. the `.v-icon` is not added by me, I think it's better to keep as original.

**Comment by pwei1018** [comment]
Temporary Url for review: https://business-edit-dev--pr-516-37q7zf1c.web.app

File: src/components/common/BcRegEntityDetails.vue
Comment (severinbeauvais) [reply src/components/common/BcRegEntityDetails.vue]: You've done things differently than I would have, and that's OK.

My initial concern -- and it is still a concern -- is that the visibility variables are in 2 components (parent and child), and I was trying to get away from that. (In fact, you have both helpToggle and isHelpVisible in ChangeBusinessType.vue.)

I think it would be more simple to just have a help component, which handles its own show/hide, that you pass a slot of help data to.

File: src/components/common/YourCompany/ChangeBusinessType.vue
Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue]: ok

Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue]: PS If you do need to set this CSS, and it's in a child component, I suggest moving this CSS to the child component instead of using deep() here.

The idea here is "try to keep related things together."

Comment (Jxio) [reply src/components/common/YourCompany/ChangeBusinessType.vue]: ok, thanks!

File: src/components/common/BcRegEntityDetails.vue
Comment (Jxio) [reply src/components/common/BcRegEntityDetails.vue]: I've replaced isHelpVisible with helpToggle. Both of these variables were already present in the existing alteration code, and I was simply utilizing them. I didn't realize initially that they were serving the same purpose. Therefore, I've now removed isHelpVisible.

Now, only ChangeBusinessType.vue is handling the show/hide logic, while BcRegEntityDetails.vue is merely retrieving the display status.

Comment (severinbeauvais) [reply src/components/common/BcRegEntityDetails.vue]: Did you try this? It seems simpler to me, and only has visibility/display status in 1 component.

> In BcRegEntityDetails.vue:
> 
> ```
> <HelpSectionToggle>
>   <template #content> 
>     
>   </template>
> </HelpSectionToggle>
> ```
> 
> In HelpSectionToggle.vue:
> 
> ```
> <template>
>   <div class="help-section-toggle">
>     <div @click="isHelpVisible= !isHelpVisible">
>       
>   </div>
> 
>   <slot name="content" v-if="isHelpVisible" />
> </template>
> ```

Comment (Jxio) [reply src/components/common/BcRegEntityDetails.vue]: I added back HelpSectionToggle.vue with slots. and only let it control visibility.

Comment (severinbeauvais) [reply src/components/common/BcRegEntityDetails.vue]: Sorry for the extra work, but it looks great! And it's clean, compact, reusable, etc. Thanks for fixing it.

Comment (Jxio) [reply src/components/common/BcRegEntityDetails.vue]: thanks for point out, it's good opportunity to learn vue slots.

**Review by severinbeauvais — APPROVED** [review]
Looks good.

**Comment by pwei1018** [comment]
Temporary Url for review: https://business-edit-dev--pr-516-37q7zf1c.web.app

**Comment by severinbeauvais** [comment]
In the future, please wait for a couple of approvals before merging. Thanks!

**Comment by Jxio** [comment]
> In the future, please wait for a couple of approvals before merging. Thanks!

ok will do it next time.

**Pull request merged**
